From 014174e874d1dc0f01dadb97eeacf86fd7c715c3 Mon Sep 17 00:00:00 2001
From: Wojciech Jablonski <wojciech.jablonski@intel.com>
Date: Tue, 5 Mar 2019 13:30:52 +0100
Subject: [PATCH 13/18] ASoC: Intel: Skl: Virt: Fix NULL ptr in pcm_close on
 SOS

In current version of pcm_close there is an assumption that DMA
buffer has been allocated, when restoring original/native DMA pages.
However some scenarios require call to  pcm_close earlier, before
DMA allocation. This patch fixes this issue by checking if DMA
buffer exist.

Change-Id: Ie0caab49b98d3519ba74db3cea8b7bbe6242dc7a
Signed-off-by: Wojciech Jablonski <wojciech.jablonski@intel.com>
Tracked-On: OAM-76768
---
 .../soc/intel/skylake/virtio/skl-virtio-be.c  | 20 ++++++++++---------
 1 file changed, 11 insertions(+), 9 deletions(-)

diff --git a/sound/soc/intel/skylake/virtio/skl-virtio-be.c b/sound/soc/intel/skylake/virtio/skl-virtio-be.c
index 374d5d7ef8bc..7f02b65f7e9b 100644
--- a/sound/soc/intel/skylake/virtio/skl-virtio-be.c
+++ b/sound/soc/intel/skylake/virtio/skl-virtio-be.c
@@ -542,17 +542,19 @@ static int vbe_skl_pcm_close(const struct skl *sdev, int vm_id,
 	int ret, cnt;
 	struct snd_pcm_substream *substream = substr_info->substream;
 	struct vfe_pcm_result *vbe_result = msg->rx_data;
-
-	const struct snd_sg_buf *sg_buf =
-			snd_pcm_substream_sgbuf(substr_info->substream);
+	struct snd_sg_buf *sg_buf;
 	u64 native_addr = substr_info->native_dma_addr;
 
-	/* restore original dma pages */
-	sg_buf->table[0].addr = native_addr;
-	native_addr &= ~(u64)0xfff;
-	for (cnt = 1; cnt < sg_buf->pages; cnt++) {
-		native_addr += PAGE_SIZE;
-		sg_buf->table[cnt].addr = native_addr;
+	if (snd_pcm_get_dma_buf(substr_info->substream)) {
+		sg_buf = snd_pcm_substream_sgbuf(substr_info->substream);
+
+		/* restore original dma pages */
+		sg_buf->table[0].addr = native_addr;
+		native_addr &= ~(u64)0xfff;
+		for (cnt = 1; cnt < sg_buf->pages; cnt++) {
+			native_addr += PAGE_SIZE;
+			sg_buf->table[cnt].addr = native_addr;
+		}
 	}
 
 	if (substr_info->pos_desc) {
-- 
2.17.1

